src/ptc/util/jms.clj

```diff
@@ -213,13 +206,16 @@
   "Throw or push to cloud at PREFIX all the files for ednified JMS message."
   [prefix jms]
   (let [workflow (get-in jms [::Properties :payload :workflow])
-        missing? (fn [k] (when (nil? (k workflow)) k))
-        check-missing (fn [k] (if (vector? k)
-                                (if (= (count (keep missing? k)) (count k)) (first k))
-                                (missing? k)))
-        missing (keep check-missing required-jms-keys)]
-    (when (seq missing)
-      (throw (IllegalArgumentException.
-              (str/join \space [missing-keys-message (sort (vec missing))]))))
-    (let [params (push-params prefix workflow)]
-      [params (push-append-to-aou-request prefix workflow params)])))
+        optional (group-by vector? required-jms-keys)
+        required (sort (optional false))
+        one-ofs  (map set (optional true))]
+    (letfn [(missing? [k] (when (nil? (k workflow)) k))
+            (none? [one-of]
+              (when (not-any? one-of (keys workflow))
+                one-of))]
+      (let [missing (concat (keep missing? required) (keep none? one-ofs))]
+        (when (seq missing)
+          (throw (IllegalArgumentException.
+                  (str/join \space [missing-keys-message (vec missing)])))))
+      (let [params (push-params prefix workflow)]
+        (push-append-to-aou-request prefix workflow params)))))
```

Comment:
I don't really get `push-append-to-aou-request` and I'm not sure changing its signature helps...
I'd feel a bit more comfortable if it did only send a request and we returned the tuple here:
```
(let [params (push-params prefix workflow)
      ptc    (str (cloud-prefix prefix workflow) "/ptc.json")]
  (push-append-to-aou-request prefix workflow params ptc)
  [params ptc])
```

Reply:
I'll call you.

Reply:
I'll break this up if you'd rather but it's only 14 lines called from one place.